lib/IntermediateRepresentation/mixin.h

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef FORTRAN_INTERMEDIATEREPRESENTATION_MIXIN_H_
+#define FORTRAN_INTERMEDIATEREPRESENTATION_MIXIN_H_
+
+#include "llvm/ADT/ilist.h"
+#include <optional>
+#include <tuple>
+#include <type_traits>
+#include <variant>
+
+namespace Fortran::IntermediateRepresentation {
+
+template<typename T, typename E = void> struct SumTypeMixin {};
+template<typename T>  // T must be std::optional<...>
+struct SumTypeMixin<T, std::enable_if_t<std::variant_size_v<T>>> {
```

Comment:
Be advised, I don't think that zero-sized variants are legal in C++.  (They do seem to work with g++ and clang, though.)

Reply:
I'm going to have to change this code a bit for clang++.